### DISCUSSION TITLE
Type Inference on Imported Symbols

### DISCUSSION THREAD

**Discussion by elbaro** [microsoft/pyright#1986] [Q&A] [answered]
According to [internals](https://github.com/microsoft/pyright/blob/main/docs/internals.md) docs,
`binder` runs on imported files to construct scope and flow graph, but `checker` does not run on them.

In the following example, how does pyright infer the type of `test2.a`?

test1.py
```py
import test2
var: str = test2.a
```

test2.py
```py
def f():
    return 1 + 2

a = f()
b: str = 'xyz'
```

```sh
❯ pyright test1.py  

test1.py
  test1.py:3:12 - error: Expression of type "int" cannot be assigned to declared type "str"
    "int" is incompatible with "str" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
Completed in 0.527sec
```

It's not even a direct assignment. This means that pyright infers the return type of `f` as well. Is this done in `binder` while walking the ast and scopes without checker or typeEvaluator?

**Comment by erictraut** [comment, marked as answer]
Are you asking simply because you're curious? Or do you see a problem here?

The binder doesn't do any type evaluation. It simply discovers symbols and associates them with scopes. It also builds the code flow graph that is used later during type evaluation.

The typeEvaluator is the component that determines the type of a symbol or an expression. It does this on demand, and it can do it in any order. It's smart enough to evaluate dependent symbol types even if that requires jumping around between files.

The checker simply visits every statement in a file and invokes the type evaluator on every statement.

In the example above, pyright is asked to analyze only one file: `test1.py`. That means it will run the checker only on that single file. It first parses and binds `test1.py`. As part of binding, it discovers that module `test2` is imported, so it also parses and binds `test2.py`.

Then the checker starts to visit each statement in `test1.py`. When it encounters the statement `var: str = test2.a`, it calls the type evaluator to evaluate the types in this statement. The type evaluator evaluates the RHS of the assignment (the `test2.a` expression). This is a "member access expression", so it first evaluates the LHS (`test2`) which evaluates to a module type. It then looks up the member named `a`, whose type is not yet known. It evaluates the type of `a` by looking at all the places where it is assigned a value. This was determined by the binder, and there's only one such assignment: `a = f()`. This is a call expression, and the LHS (`f`) now needs to be evaluated. There is a function declaration for `f`, but it has no declared return type, so now the type evaluator infers the return type of `f` based on all of the return statements in the function's body. (The return statements were identified at binding time.) In your example, there is only one return statement, and it's returning an expression whose type is inferred as `int`. So the inferred return type of `f` is `int`. That means the inferred type of `a` is also `int`.

Hopefully that gives you a sense for how the various components work. Let me know if any of that was unclear.

**Reply by elbaro** [reply]
Thanks for the great answer, it is very clear now.

The question was for curiosity.
Pyright docs are well written and I'm learning a lot from it.
